### PR TITLE
Add mixin for smb pipe auditing

### PR DESCRIPTION
**Pull request by Auxilus**
PoW

```
msf5 auxiliary(scanner/smb/smb_ms17_010) > run

[+] 192.168.0.2:445       - Host is likely VULNERABLE to MS17-010! - Windows 7 Ultimate 7601 Service Pack 1 x64 (64-bit)
[*] 192.168.0.2:445       - Checking for accessible named pipes
[+] 192.168.0.2:445       - Found accessible named pipe: netlogon
[+] 192.168.0.2:445       - Found accessible named pipe: lsarpc
[+] 192.168.0.2:445       - Found accessible named pipe: samr
[+] 192.168.0.2:445       - Found accessible named pipe: browser
[+] 192.168.0.2:445       - Found accessible named pipe: atsvc
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
```

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use auxiliary/scanner/smb/smb_ms17_010`
- [x] set RHOSTS
- [x] run

## TODO
- [x] Add scanner module for pipe auditing
- [x] Fix connection issues
- [x] Add named pipes wordlist
- [ ] Sort pipe wordlist

### CODE REVIEW

**Comment by wvu-r7** [comment]
This looks a lot like `pipe_auditor`. Consider consolidating the functionality into one place?

**Comment by Auxilus** [comment]
You'd have to walk me through if you don't mind , sorry for the trouble

**Comment by Auxilus** [comment]
I fount this pipe scanner code in ms17_010_eternalblue , copy pasted it to the scanner , and it worked....

**Comment by wvu-r7** [comment]
No worries. This was one of our to-dos from #9473.

**Comment by Auxilus** [comment]
Yes ik :)

**Comment by wvu-r7** [comment]
You mean `lib/msf/core/exploit/smb/client/psexec_ms17_010.rb`? It was copied from `pipe_auditor`, IIRC.

What we should do is move the code into a general-purpose library and also add the ability to search for additional pipes, then rework the modules to use the consolidated code.

cc @zerosum0x0

**Comment by Auxilus** [comment]
you mean adding it to `lib/msf/core/auxiliary` then include it where ever its needed?

**Comment by wvu-r7** [comment]
Probably a good idea to create `data/wordlists/named_pipes.txt`, too.

**Comment by Auxilus** [comment]
Yeah,I can work on it tomorrow , its 4 am :)

**Comment by Auxilus** [comment]
Looks like scanner should return more than one accessible named pipes, so do we have to modify `psexec_ms17_010.rb`?

Or just return the first accessible pipe name rather than returning array?

File: modules/auxiliary/scanner/smb/smb_ms17_010.rb [hunk 1]
```diff
@@ -90,6 +90,43 @@ def run_host(ip)
         end
 
         print_good("Host is likely VULNERABLE to MS17-010! - #{os}")
+        
+        
+        # Detect accessible named pipes
+        print_status("Checking for accessible named pipes")
```
Comment (zerosum0x0) [inline modules/auxiliary/scanner/smb/smb_ms17_010.rb]: make this print a vprint_status imo

File: modules/auxiliary/scanner/smb/smb_ms17_010.rb [hunk 2]
```diff
@@ -90,6 +90,43 @@ def run_host(ip)
...
+                'W32TIME_ALT',
+                'wkssvc',
+                'PIPE_EVENTROOT\CIMV2SCM EVENT PROVIDER',
+                'db2remotecmd'
+        ]
+
+        target_pipes.each do |pipe|
+                pipe_name = "#{pipe}"
+                pipe_handle = self.simple.create_pipe(pipe_name, 'o')
+                print_good("Found accessible named pipe: #{pipe}")
```
Comment (zerosum0x0) [inline modules/auxiliary/scanner/smb/smb_ms17_010.rb]: You can realistically end after the first named pipe here. But if you want to list them all, I'd comma separate them on a single line.

File: modules/auxiliary/scanner/smb/smb_ms17_010.rb [hunk 3]
```diff
@@ -90,6 +90,43 @@ def run_host(ip)
...
+        end
```
Comment (zerosum0x0) [inline modules/auxiliary/scanner/smb/smb_ms17_010.rb]: maybe a vprint_status that no named pipe was found

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by Auxilus** [comment]
I've added rough code for the idea above

**Comment by Auxilus** [comment]
There is a connection issue that I can't find origin of... 
screenshot_20180224-111444

**Comment by Auxilus** [comment]
I'm getting `auxilus@termux:~/g/metasploit-framework > ./msfconsole                  ^[[A/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/core/exploit/smb/client/psexec_ms17_010.rb:7:in `<module:Psexec_MS17_010>': uninitialized constant Msf::Auxiliary::PIPEAudit (NameError)`

**Comment by jmartin-r7** [comment]
Travis failures are for failing checks in `msftidy.rb`.   Please run locally and address any issues.

```
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/auxiliary/scanner/smb/smb_ms17_010.rb:93 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/smb/smb_ms17_010.rb:94 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/smb/smb_ms17_010.rb:130 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/smb/smb_ms17_010.rb:133 - [WARNING] Spaces at EOL
modules/auxiliary/scanner/smb/smb_ms17_010.rb:139 - [WARNING] Spaces at EOL
```

**Comment by Auxilus** [comment]
The error I'm getting is
```
msf5 exploit(windows/smb/ms17_010_psexec) > exploit

[*] Started reverse TCP handler on 192.168.0.4:4444
[*] 192.168.0.5:445 - Target OS: Windows 7 Ultimate 7601 Service Pack 1
[-] 192.168.0.5:445 - NoMethodError
[-] 192.168.0.5:445 - undefined method `connect_to_pipe' for #<Msf::Modules::Mod6578706c6f69742f77696e646f77732f736d622f6d7331375f3031305f707365786563::MetasploitModule:0x0000007f8826db00>
Did you mean?  connect_timeout
[-] 192.168.0.5:445 - /data/data/com.termux/files/home/git/metasploit-framework/lib/msf/core/exploit/smb/client/psexec_ms17_010.rb:335:in `find_accessible_named_pipe'
/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/core/exploit/smb/client/psexec_ms17_010.rb:37:in `eternal_pwn'
/data/data/com.termux/files/home/git/metasploit-framework/modules/exploits/windows/smb/ms17_010_psexec.rb:99:in `exploit'
/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/core/exploit_driver.rb:206:in `job_run_proc'
/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/core/exploit_driver.rb:167:in `run'
/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/base/simple/exploit.rb:136:in `exploit_simple'
/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/base/simple/exploit.rb:161:in `exploit_simple'
/data/data/com.termux/files/home/git/metasploit-framework/lib/msf/ui/console/command_dispatcher/exploit.rb:110:in `cmd_exploit'
/data/data/com.termux/files/home/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:548:in `run_command'
/data/data/com.termux/files/home/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:510:in `block in run_single'
/data/data/com.termux/files/home/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:504:in `each'
/data/data/com.termux/files/home/git/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:504:in `run_single'
/data/data/com.termux/files/home/git/metasploit-framework/lib/rex/ui/text/shell.rb:208:in `run'
/data/data/com.termux/files/home/git/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/data/data/com.termux/files/home/git/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:48:in `<main>'
[*] Exploit completed, but no session was created.
```




Error is same when I run in Ubuntu

**Comment by Auxilus** [comment]
```
auxilus@termux:~/d/metasploit-framework > ./msfconsole
/data/data/com.termux/files/home/doc/metasploit-framework/lib/msf/core/exploit/smb/client/psexec_ms17_010.rb:6:in `<module:Psexec_MS17_010>': uninitialized constant Msf::Auxiliary::PipeAudit (NameError)
```

**Comment by Auxilus** [comment]
The changes are working now

**Comment by Auxilus** [comment]
@wvu-r7 can you take a look at scanner module? Gives me connection reset error
Also see if other changes are needed. Thanks.

**Comment by busterb** [comment]
@bwatters-r7 / @jmartin-r7 could we get a retest of the related modules here?

**Comment by wvu-r7** [comment]
558ecbcb4993776d1dfe6541167912de7eca4209
23a7444646f1829eb6aa9c2b25bd00c9551aa90d
e4c026fffd73183a53c908ffc40b977257311e43
09cb4a52df1b822f3040990ddb56e49e3bc9728c

**Comment by wvu-r7** [comment]
# Release Notes

This adds improvements to pipe auditing both to Framework core and the MS17-010 modules.

**Comment by busterb** [comment]
👍 nice work!

**Comment by Auxilus** [comment]
Thanks for the help @wvu-r7 @jmartin-r7 @busterb

**Comment by Auxilus** [comment]
also @zerosum0x0

**Comment by wvu-r7** [comment]
You got it! Thanks for spearheading this!

**Comment by faizzaki212** [comment]
i scanned a host using auxiliary/scanner/smb/smb_ms17_010 then it says [+] host is likely vulnerable to ms17! but when i use all of the ms17 exploit, it doesn't work. my target is windows 7 ultimate pro and i'm running mac OSX sierra

**Comment by wvu-r7** [comment]
Please don't comment on old pull requests for support. Check this out: https://metasploit.com/help.

**Pull request merged**